### DISCUSSION TITLE
Access Modifiers

### DISCUSSION THREAD

**Discussion by tscottjendev** [microsoft/alguidelines#210] [BC Best Practices]
What is the best practice for access modifiers?  Should all (eligible) objects require it like for the [LC011](https://github.com/StefanMaron/BusinessCentral.LinterCop/wiki/LC0011)?

Should public and internal procedures be separated into separate objects?

Should all procedures in an internal object, have the internal modifier?

**Comment by tscottjendev** [comment]
Should the default position be Internal unless there is a specific reason not to be?

What about events in an internal object?  They should be marked as `[InternalEvent]` or moved to a public codeunit I guess.

Thoughts?

**Reply by waldo1001** [reply]
Default internal makes sense to me.  Doing that though, will close down your apps by default, and I'm not sure our consultants/developers would have liked that (we didn't do that ;-)), because just imagine: when you are implementing your own product, which you release like every month - you'd even have to wait a month to "open" functionality when necessary to be able to implement your product at a customer.
The other way around: when access is public by default, you'll obviously have the reverse effect: you'll be very efficient, but run into unstable situations just because you might extend or use stuff that wasn't supposed to be used 🤷‍♀️.

hard to go for a default .. 🤔

**Reply by StefanMaron** [reply]
My approach to this is, to make "normal" objects internal by default. Because in most cases those objects to not fulfill requirements an open "API" would have. Documentation, no braking changes, versions and possibly more. And for my understanding, all public objects should be treated as an API.

But I know that sometimes it is needed to have some hotfixes deployed or customer individual customizations made that cant be achieved by the API only. This is especially a problem for AppSource Apps.

To solve this I, sometimes created a "[Appname] Customizations" extension, which was added to the internal visibles to property. If multiple customers needed individual customizations, all of them would have the app created for them with their changes and deployed as PTE. Even though the App ID and App Name were the same, the customer might get different customizations based on their needs.
